File: products/mcp_server/src/a2a_mcp_server/gateway_client.py
"""Thin async HTTP client for the A2A Commerce Gateway.

Only exposes what the MCP server needs:

  * ``list_tools()`` — fetch the full tool catalogue from ``GET /v1/pricing``
  * ``invoke_tool(name, params)`` — execute a single tool via ``POST /v1/batch``

``/v1/batch`` is used rather than the legacy ``/v1/execute`` endpoint because
the latter is restricted to connector tools in v1.2+, while batch accepts
every tool registered in the gateway's ``TOOL_REGISTRY``.
"""

from __future__ import annotations

from typing import Any

import httpx

from a2a_mcp_server._version import __version__

_DEFAULT_TIMEOUT = httpx.Timeout(connect=5.0, read=30.0, write=10.0, pool=5.0)


class GatewayError(RuntimeError):
    """Raised when the gateway returns an error result."""


class GatewayAuthError(GatewayError):
    """Raised when the gateway rejects the API key (401/403)."""


class GatewayRateLimitError(GatewayError):
    """Raised when the gateway returns 429 Too Many Requests."""


class GatewayClient:
    """Async HTTP client for the A2A gateway, scoped to MCP server needs."""

    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        timeout: httpx.Timeout | float | None = None,
        transport: httpx.AsyncBaseTransport | None = None,
        user_agent: str | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._client = httpx.AsyncClient(
            base_url=self._base_url,
            timeout=timeout or _DEFAULT_TIMEOUT,
            transport=transport,
            headers={
                "Authorization": f"Bearer {api_key}",
                "User-Agent": user_agent or f"a2a-mcp-server/{__version__}",
                "Accept": "application/json",
            },
        )

    async def close(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> GatewayClient:
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        await self.close()

    async def list_tools(self) -> list[dict[str, Any]]:
        """GET /v1/pricing → list of catalog tool definitions."""
        resp = await self._client.get("/v1/pricing")
        _raise_for_status(resp)
        body = resp.json()
        tools = body.get("tools") if isinstance(body, dict) else None
        if not isinstance(tools, list):
            raise GatewayError(f"Unexpected /v1/pricing payload shape: {body!r}")
        return tools

    async def invoke_tool(self, name: str, params: dict[str, Any] | None = None) -> Any:
        """POST /v1/batch with a single call — unwraps and returns the result."""
        payload = {"calls": [{"tool": name, "params": params or {}}]}
        resp = await self._client.post("/v1/batch", json=payload)
        _raise_for_status(resp)
        body = resp.json()
        results = body.get("results") if isinstance(body, dict) else None
        if not results or not isinstance(results, list):
            raise GatewayError(f"Unexpected /v1/batch payload shape: {body!r}")
        first = results[0]
        if first.get("success"):
            return first.get("result")
        err = first.get("error") or {}
        code = err.get("code", "gateway_error")
        message = err.get("message", "Unknown gateway error")
        raise GatewayError(f"[{code}] {message}")


def _raise_for_status(resp: httpx.Response) -> None:
    if resp.status_code == 401 or resp.status_code == 403:
        raise GatewayAuthError(_extract_message(resp) or "Gateway rejected API key")
    if resp.status_code == 429:
        raise GatewayRateLimitError(_extract_message(resp) or "Gateway rate limit exceeded")
    if resp.status_code >= 400:
        raise GatewayError(f"Gateway returned {resp.status_code}: {_extract_message(resp) or resp.text}")


def _extract_message(resp: httpx.Response) -> str | None:
    try:
        body = resp.json()
    except ValueError:
        return None
    if isinstance(body, dict):
        err = body.get("error")
        if isinstance(err, dict):
            return err.get("message")
        if isinstance(body.get("detail"), str):
            return body["detail"]
    return None

File: sdk-ts/mcp-server/src/gatewayClient.ts
/**
 * Thin async HTTP client for the A2A Commerce Gateway.
 *
 * Only exposes what the MCP server needs:
 *   - listTools()              → GET /v1/pricing
 *   - invokeTool(name, params) → POST /v1/batch (single call)
 *
 * /v1/batch is used instead of the legacy /v1/execute endpoint because the
 * latter is restricted to connector tools in gateway v1.2+.
 */

export class GatewayError extends Error {
  readonly code: string;
  constructor(code: string, message: string) {
    super(message);
    this.name = 'GatewayError';
    this.code = code;
  }
}

export class GatewayAuthError extends GatewayError {
  constructor(message: string) {
    super('unauthorized', message);
    this.name = 'GatewayAuthError';
  }
}

export class GatewayRateLimitError extends GatewayError {
  constructor(message: string) {
    super('rate_limit_exceeded', message);
    this.name = 'GatewayRateLimitError';
  }
}

export interface GatewayClientOptions {
  baseUrl: string;
  apiKey: string;
  userAgent?: string;
  fetch?: typeof fetch;
}

export interface CatalogTool {
  name: string;
  service?: string;
  description?: string;
  input_schema?: Record<string, unknown>;
  output_schema?: Record<string, unknown>;
  pricing?: { per_call?: number };
  tier_required?: string;
  [key: string]: unknown;
}

export class GatewayClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly userAgent: string;
  private readonly fetchImpl: typeof fetch;

  constructor(options: GatewayClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, '');
    this.apiKey = options.apiKey;
    this.userAgent = options.userAgent ?? '@greenhelix/mcp-server/0.1.0';
    this.fetchImpl = options.fetch ?? globalThis.fetch;
    if (!this.fetchImpl) {
      throw new Error('No global fetch available. Node >=18 is required, or inject options.fetch.');
    }
  }

  async listTools(): Promise<CatalogTool[]> {
    const resp = await this.fetchImpl(`${this.baseUrl}/v1/pricing`, {
      method: 'GET',
      headers: this.headers(),
    });
    await this.ensureOk(resp);
    const body = (await resp.json()) as { tools?: CatalogTool[] };
    if (!body || !Array.isArray(body.tools)) {
      throw new GatewayError('bad_response', `Unexpected /v1/pricing payload: ${JSON.stringify(body)}`);
    }
    return body.tools;
  }

  async invokeTool(name: string, params: Record<string, unknown> = {}): Promise<unknown> {
    const resp = await this.fetchImpl(`${this.baseUrl}/v1/batch`, {
      method: 'POST',
      headers: {
        ...this.headers(),
        'content-type': 'application/json',
      },
      body: JSON.stringify({ calls: [{ tool: name, params }] }),
    });
    await this.ensureOk(resp);
    const body = (await resp.json()) as { results?: Array<{ success: boolean; result?: unknown; error?: { code?: string; message?: string } }> };
    if (!body || !Array.isArray(body.results) || body.results.length === 0) {
      throw new GatewayError('bad_response', `Unexpected /v1/batch payload: ${JSON.stringify(body)}`);
    }
    const first = body.results[0];
    if (first.success) {
      return first.result;
    }
    const code = first.error?.code ?? 'gateway_error';
    const message = first.error?.message ?? 'Unknown gateway error';
    throw new GatewayError(code, `[${code}] ${message}`);
  }

  private headers(): Record<string, string> {
    return {
      authorization: `Bearer ${this.apiKey}`,
      'user-agent': this.userAgent,
      accept: 'application/json',
    };
  }

  private async ensureOk(resp: Response): Promise<void> {
    if (resp.ok) return;
    const message = await extractMessage(resp);
    if (resp.status === 401 || resp.status === 403) {
      throw new GatewayAuthError(message ?? 'Gateway rejected API key');
    }
    if (resp.status === 429) {
      throw new GatewayRateLimitError(message ?? 'Gateway rate limit exceeded');
    }
    throw new GatewayError('http_error', `Gateway returned ${resp.status}: ${message ?? ''}`);
  }
}

async function extractMessage(resp: Response): Promise<string | null> {
  try {
    const body = await resp.json();
    if (body && typeof body === 'object') {
      const err = (body as Record<string, unknown>).error;
      if (err && typeof err === 'object' && typeof (err as Record<string, unknown>).message === 'string') {
        return (err as Record<string, string>).message;
      }
      if (typeof (body as Record<string, unknown>).detail === 'string') {
        return (body as Record<string, string>).detail;
      }
    }
  } catch {
    /* ignore */
  }
  return null;
}
